### ISSUE TITLE
Fused cross entropy loss

### ISSUE THREAD

**Issue by lyprince** [pytorch/torchtitan#1076]
I'd like to fuse the final linear projection and the cross entropy loss (e.g., with https://github.com/apple/ml-cross-entropy/) since there is a huge memory spike caused by needing to store the `[batch x vocab]` when computing the loss.

`TrainSpec` at the moment does not build a `loss_fn`, simply sets it as an attribute. For a fused loss it would be necessary to build a `loss_fn` using the parameters from the final linear projection of the model.

I would be happy to implement a `build_loss_fn` as part of TrainSpec and submit a PR.

**Comment by lyprince** [comment]
Ah - I was looking at our old fork. Seen this has already been done.
